racket/collects/racket/private/list.rkt

```diff
@@ -323,31 +323,36 @@
             [else (cons (fcn j)
                         (recr (add1 j) (sub1 i)))])))
 
-  (define-values [compose1 compose]
+  (define-values (compose1 compose)
     (let ()
+      (define (id? f) (equal? values f))
```

Comment:
I think this needs to be `eq?` instead of `equal?` in case a chaperoned `values` is supplied.

Reply:
If `eq?` is used here I am concerned that TR may behave unexpectedly.

For example:
```racket
#lang typed/racket/base
(module m racket/base
  (provide (all-defined-out))
  (define (id? f) (eq? values f)))
(require/typed 'm
  [id? (-> Any Boolean)])
(displayln (id? values)) ; #f

```

Reply:
That seems like the behavior we would expect in that situation.

Reply:
Yes, so I think `equal?` might be appropriate here.

Reply:
What's the motivation for this PR? I personally don't think that it's worth the complexity added to the code, unless you can demonstrate that it makes the function more efficient, and this programming pattern is frequently seen in the wild.

Trying to ad-hoc recognize "identity functions" is bound to fail. You can test for `values`, and perhaps chaperoned `values`. Currently you are missing `identity` from `racket/function`. And what about `(lambda (x) x)`? What about `(lambda xs (apply values xs))`, etc, etc.

Reply:
I guess it's worth it, the reason why `compose` should ignore `values` is because `(compose)` evaluates to `values`, so I'm not going to consider other cases.

Treating `compose` as monoid may be convenient for some scenarios (I guess `Qi` may need it).

Reply:
dropping a chaperoned version of values is absolutely not the right thing to do. That could cause a check that should happen not to happen.

Reply:
You are right, I'll fix it later.